# DESAFIOS SEMANA 05

### Desafio 1 Aula:
Vendas mensais de uma loja de roupas. Arquivo: dataset_squad_b_loja_roupas.csv

Faça uma análise das vendas por categoria e mês .
**Comando:** rode `head()`, `describe()` e pelo menos um `groupby()` e apresente 2 conclusões reais que ajudem a diretoria a decidir o mix de compras.

In [1]:
import pandas as pd
from IPython.display import display

dataset_squad_b = pd.read_csv("dataset/dataset_squad_b_loja_roupas.csv")

display(dataset_squad_b.head())
display(dataset_squad_b.describe())
display(dataset_squad_b.groupby("categoria")["vendas"].sum().sort_values(ascending=False))
display(dataset_squad_b.groupby("mes")["vendas"].sum())

# Conclusão 1: "Camisas" é a categoria mais vendida no período (760 unidades), quase o dobro de "Casacos" (320) — vale reforçar estoque de camisas.
# Conclusão 2: Março foi o mês mais fraco (320 unidades) e Fevereiro o mais forte (400) — vale investigar o que mudou em março (promoção? sazonalidade?).

,mes,categoria,vendas
0,Jan,Camisas,120
1,Fev,Camisas,150
2,Mar,Camisas,90
3,Abr,Camisas,130
4,Mai,Camisas,110


,vendas
count,24.000000
mean,89.166667
std,40.234818
min,15.000000
25%,60.000000
50%,92.500000
75%,112.500000
max,160.000000


categoria
Camisas     760
Calcas      555
Vestidos    505
Casacos     320
Name: vendas, dtype: int64

mes
Abr    365
Fev    400
Jan    335
Jun    355
Mai    365
Mar    320
Name: vendas, dtype: int64

### Desafio 2 Aula:
Assinantes de streaming por região e plano. Arquivo: dataset_squad_c_streaming.csv

Faça um mapeamento de onde estão os assinantes hoje e quais planos vendem mais em cada região.
**Comando:** rode `head()`, `describe()` e pelo menos um `groupby()` e apresente 2 conclusões reais que ajudem a decidir onde investir.


In [2]:
import pandas as pd
from IPython.display import display

dataset_squad_c = pd.read_csv("dataset/dataset_squad_c_streaming.csv")

display(dataset_squad_c.head())
display(dataset_squad_c.describe())
display(dataset_squad_c.groupby("regiao")["assinantes"].sum().sort_values(ascending=False))
display(dataset_squad_c.groupby("plano")["assinantes"].sum().sort_values(ascending=False))

# Conclusão 1: a região Sudeste concentra quase 40% de todos os assinantes (28.200 de ~72.100) — a região Norte tem só 2.900, o menor volume.
# Conclusão 2: o plano Básico é o mais assinado (25.400), à frente de Premium (18.300) e Família (11.900) — sugere que preço pesa mais que recursos extras para a maioria dos assinantes.

,regiao,plano,assinantes
0,Sudeste,Basico,12000
1,Sudeste,Premium,9800
2,Sudeste,Familia,6400
3,Sul,Basico,4500
4,Sul,Premium,3100


,assinantes
count,15.000000
mean,3706.666667
std,3392.526801
min,500.000000
25%,1500.000000
50%,2200.000000
75%,4850.000000
max,12000.000000


regiao
Sudeste         28200
Nordeste        10300
Sul              9800
Centro-Oeste     4400
Norte            2900
Name: assinantes, dtype: int64

plano
Basico     25400
Premium    18300
Familia    11900
Name: assinantes, dtype: int64

### Desafio 3 Aula
Avaliações de um app. Arquivo: dataset_squad_d_app.csv

Faça uma análise para saber quais categorias de app têm melhor avaliação e mais downloads.
**Comando:** rode `head()`, `describe()` e pelo menos um `groupby()` e apresente 2 conclusões reais que ajudem a escolher a categoria do próximo lançamento.

In [3]:
import pandas as pd
from IPython.display import display

dataset_squad_d = pd.read_csv("dataset/dataset_squad_d_app.csv")

display(dataset_squad_d.head())
display(dataset_squad_d.describe())
display(dataset_squad_d.groupby("categoria_app")[["nota", "downloads_mil"]].mean().sort_values("nota", ascending=False))

# Conclusão 1: "Educação" tem a melhor nota média (4.65), mas é a categoria com MENOS downloads (81 mil) — nota alta não garante volume.
# Conclusão 2: "Jogos" tem a pior nota média (4.03) e, ainda assim, de longe o maior volume de downloads (532 mil) — o próximo lançamento precisa decidir entre "nota" e "alcance" como prioridade.

,categoria_app,nota,downloads_mil
0,Jogos,4.2,500
1,Jogos,3.9,800
2,Jogos,4.5,620
3,Jogos,3.5,210
4,Produtividade,4.7,120


,nota,downloads_mil
count,15.000000,15.000000
mean,4.386667,244.000000
std,0.388893,225.580204
min,3.500000,45.000000
25%,4.150000,92.500000
50%,4.500000,150.000000
75%,4.650000,280.000000
max,4.900000,800.000000


,nota,downloads_mil
categoria_app,,
Educacao,4.650000,81.25
Produtividade,4.525000,188.75
Saude,4.333333,150.00
Jogos,4.025000,532.50


### Desafio 4
Atendimentos de uma clínica. Arquivo: dataset_squad_e_clinica.csv

Faça uma análise para entender se o tempo de espera está afetando a satisfação dos pacientes, especialidade por especialidade, antes de decidir se contrata mais profissionais.

**Comando:** rode `head()`, `describe()` e pelo menos um `groupby()` e apresente 2 conclusões reais que ajudem a decidir sobre a contratação.

In [4]:
import pandas as pd
from IPython.display import display

dataset_squad_e = pd.read_csv("dataset/dataset_squad_e_clinica.csv")

display(dataset_squad_e.head())
display(dataset_squad_e.describe())
display(dataset_squad_e.groupby("especialidade")[["tempo_espera_min", "satisfacao"]].mean().sort_values("tempo_espera_min", ascending=False))

# Conclusão 1: Ortopedia tem o maior tempo médio de espera (39 min) e a menor satisfação média (3.2) entre as especialidades.
# Conclusão 2: Pediatria tem o menor tempo de espera (18.75 min) e a maior satisfação (4.75) — os dados sugerem que tempo de espera e satisfação andam juntos, o que apoiaria reforçar equipe onde a espera é maior.

,especialidade,tempo_espera_min,satisfacao
0,Clinico Geral,25,4
1,Clinico Geral,40,3
2,Clinico Geral,15,5
3,Clinico Geral,55,2
4,Clinico Geral,20,5


,tempo_espera_min,satisfacao
count,17.000000,17.00000
mean,28.058824,4.00000
std,14.669145,1.06066
min,10.000000,2.00000
25%,18.000000,3.00000
50%,25.000000,4.00000
75%,35.000000,5.00000
max,60.000000,5.00000


,tempo_espera_min,satisfacao
especialidade,,
Ortopedia,39.000000,3.200000
Clinico Geral,31.000000,3.800000
Pediatria,18.750000,4.750000
Dermatologia,17.333333,4.666667


## Desafio 1 - Ler arquivos

In [11]:
import pandas as pd

# ====================
# LEITURA BASES DATASET
df_loja_roupas = pd.read_csv("dataset/dataset_squad_b_loja_roupas.csv")
df_streaming = pd.read_csv("dataset/dataset_squad_c_streaming.csv")
df_app = pd.read_csv("dataset/dataset_squad_d_app.csv")
df_clinica = pd.read_csv("dataset/dataset_squad_e_clinica.csv")
df_vendas_papelaria = pd.read_csv("dataset/vendas_papelaria.csv")


# ====================
# EXIBIÇÃO BASES
print("dataset_squad_b_loja_roupas")
display(df_loja_roupas)

print()
print("dataset_squad_c_streaming")
display(df_streaming)

print()
print("dataset/dataset_squad_d_app")
display(df_app)

print()
print("dataset/dataset_squad_e_clinica")
display(df_clinica)

print()
print("dataset/vendas_papelaria")
display(df_vendas_papelaria)


dataset_squad_b_loja_roupas


,mes,categoria,vendas
0,Jan,Camisas,120
1,Fev,Camisas,150
2,Mar,Camisas,90
3,Abr,Camisas,130
4,Mai,Camisas,110
5,Jun,Camisas,160
6,Jan,Calcas,80
7,Fev,Calcas,95
8,Mar,Calcas,100
9,Abr,Calcas,85



dataset_squad_c_streaming


,regiao,plano,assinantes
0,Sudeste,Basico,12000
1,Sudeste,Premium,9800
2,Sudeste,Familia,6400
3,Sul,Basico,4500
4,Sul,Premium,3100
5,Sul,Familia,2200
6,Nordeste,Basico,5200
7,Nordeste,Premium,3200
8,Nordeste,Familia,1900
9,Centro-Oeste,Basico,2100



dataset/dataset_squad_d_app


,categoria_app,nota,downloads_mil
0,Jogos,4.2,500
1,Jogos,3.9,800
2,Jogos,4.5,620
3,Jogos,3.5,210
4,Produtividade,4.7,120
5,Produtividade,4.5,200
6,Produtividade,4.8,95
7,Produtividade,4.1,340
8,Educacao,4.8,60
9,Educacao,4.6,90



dataset/dataset_squad_e_clinica


,especialidade,tempo_espera_min,satisfacao
0,Clinico Geral,25,4
1,Clinico Geral,40,3
2,Clinico Geral,15,5
3,Clinico Geral,55,2
4,Clinico Geral,20,5
5,Pediatria,15,5
6,Pediatria,20,5
7,Pediatria,10,5
8,Pediatria,30,4
9,Ortopedia,30,4



dataset/vendas_papelaria


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45
2,Caderno,Papelaria,Mar,12.5,28
3,Caderno,Papelaria,Abr,12.5,50
4,Lapis,Papelaria,Jan,2.0,80
5,Lapis,Papelaria,Fev,2.0,70
6,Lapis,Papelaria,Mar,2.0,90
7,Lapis,Papelaria,Abr,2.0,65
8,Caneta,Papelaria,Jan,3.5,60
9,Caneta,Papelaria,Fev,3.5,55


In [17]:
# ====================
# df.head()
print("df.head()") 
print("Mostrar as 5 primeiras linhas do DataFrame")
display(df_vendas_papelaria.head())

print()
print("Também é possível escolher a quantidade entre parentes()")
display(df_vendas_papelaria.head(2))

df.head()
Mostrar as 5 primeiras linhas do DataFrame


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45
2,Caderno,Papelaria,Mar,12.5,28
3,Caderno,Papelaria,Abr,12.5,50
4,Lapis,Papelaria,Jan,2.0,80



Também é possível escolher a quantidade entre parentes()


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45


In [18]:
# ====================
# df.info()
print("df.info()")
print("Ele responde:\n",
      "Quantas linhas existem?\n",
      "Quantas colunas?\n",
      "Qual o tipo de cada coluna?\n",
      "Existem valores vazios?\n")
display(df_vendas_papelaria.info())

df.info()
Ele responde:
 Quantas linhas existem?
 Quantas colunas?
 Qual o tipo de cada coluna?
 Existem valores vazios?

<class 'pandas.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   produto             24 non-null     str    
 1   categoria           24 non-null     str    
 2   mes                 24 non-null     str    
 3   preco_unitario      24 non-null     float64
 4   quantidade_vendida  24 non-null     int64  
dtypes: float64(1), int64(1), str(3)
memory usage: 1.1 KB


None

In [19]:
# ====================
# df.describe()
print("df.describe()")
print("Esse comando faz calculos de estatísticas automáticas nas colunas numéricas.")
print("count = Quantidade de valores.\n",
      "mean = Média.\n",
      "std = Desvio padrão. Quanto os dados variam.\n",
      "min = Menor valor.\n",
      "max = Maior valor.\n",
      "50% = Mediana.\n",
      "25% = Primeiro quartil.\n",
      "75% = Terceiro quartil.\n\n")
display(df_vendas_papelaria.describe())

df.describe()
Esse comando faz calculos de estatísticas automáticas nas colunas numéricas.
count = Quantidade de valores.
 mean = Média.
 std = Desvio padrão. Quanto os dados variam.
 min = Menor valor.
 max = Maior valor.
 50% = Mediana.
 25% = Primeiro quartil.
 75% = Terceiro quartil.




,preco_unitario,quantidade_vendida
count,24.000000,24.000000
mean,34.833333,35.458333
std,53.816651,26.663734
min,2.000000,4.000000
25%,3.500000,11.750000
50%,9.250000,26.500000
75%,35.000000,56.250000
max,150.000000,90.000000


In [20]:
# ====================
# # df.shape
print("df.shape")
print("Mostra a quantidade de (linhas, colunas)")
print(df_vendas_papelaria.shape)

print()
print("Pode pegar separadamente.")
print("df_csv.shape[0] = linhas")
print(df_vendas_papelaria.shape[0])

print()
print("df_csv.shape[1] = colunas")
display(df_vendas_papelaria.shape[1])

df.shape
Mostra a quantidade de (linhas, colunas)
(24, 5)

Pode pegar separadamente.
df_csv.shape[0] = linhas
24

df_csv.shape[1] = colunas


5

In [21]:
# ====================
# df.columns
print()
print("df.columns")
print("Mostra apenas os nomes das colunas.")
print(df_vendas_papelaria.columns)


df.columns
Mostra apenas os nomes das colunas.
Index(['produto', 'categoria', 'mes', 'preco_unitario', 'quantidade_vendida'], dtype='str')


## Desafio 2 - Selecionar dados

In [25]:
import pandas as pd

# ====================
# LEITURA ARQUIVO CSV
df_vendas_papelaria = pd.read_csv("dataset/vendas_papelaria.csv")


# ====================
# VISUALIZAÇÃO
display(df_vendas_papelaria.head(2))

,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45


In [26]:
# ====================
# NOMES DAS COLUNAS
print("Mostrar os nomes das colunas:")
display(df_vendas_papelaria.columns) 

Mostrar os nomes das colunas:


Index(['produto', 'categoria', 'mes', 'preco_unitario', 'quantidade_vendida'], dtype='str')

In [27]:
# ====================
# FILTRAR UM COLUNA
print("\n")
print("Mostrar apenas os resultados da coluna 'categoria':")
print(df_vendas_papelaria["categoria"])



Mostrar apenas os resultados da coluna 'categoria':
0      Papelaria
1      Papelaria
2      Papelaria
3      Papelaria
4      Papelaria
5      Papelaria
6      Papelaria
7      Papelaria
8      Papelaria
9      Papelaria
10     Papelaria
11     Papelaria
12     Papelaria
13     Papelaria
14     Papelaria
15     Papelaria
16    Acessorios
17    Acessorios
18    Acessorios
19    Acessorios
20    Acessorios
21    Acessorios
22    Acessorios
23    Acessorios
Name: categoria, dtype: str


In [29]:
# ====================
# FILTRAR DUAS COLUNAS
print("Mostrar os resultadas das colunas 'categoria' e 'mes':")
display(df_vendas_papelaria[["categoria","mes"]])

Mostrar os resultadas das colunas 'categoria' e 'mes':


,categoria,mes
0,Papelaria,Jan
1,Papelaria,Fev
2,Papelaria,Mar
3,Papelaria,Abr
4,Papelaria,Jan
5,Papelaria,Fev
6,Papelaria,Mar
7,Papelaria,Abr
8,Papelaria,Jan
9,Papelaria,Fev


In [ ]:
# ====================
# df.iloc() - Mostrar resultado conforme indice
print("Mostrar o resultado pela posição (índice numérico)")
display(df_vendas_papelaria.iloc[0])

Mostrar o resultado pela posição (índice numérico)


produto                 Caderno
categoria             Papelaria
mes                         Jan
preco_unitario             12.5
quantidade_vendida           30
Name: 0, dtype: object

In [ ]:
# ====================
# .loc() 
print("Mostrar o resultado pela posição (índice + nome do índice)")
display(df_vendas_papelaria.loc[0,"categoria"])

Mostrar o resultado pela posição (índice + nome do índice)


'Papelaria'

In [42]:
# ====================
# Filtros
print("Filtro na coluna 'categoria'")
display(df_vendas_papelaria[df_vendas_papelaria["categoria"]=="Papelaria"])

print()
print("Filtro na coluna 'produto'")
display(df_vendas_papelaria[df_vendas_papelaria["produto"]=="Caderno"])

print()
print("Filtro nas colunas 'categoria' e 'produto'")
display(
    df_vendas_papelaria[
        (df_vendas_papelaria["produto"]=="Caderno") & 
        (df_vendas_papelaria["categoria"]=="Papelaria")
    ]
)

Filtro na coluna 'categoria'


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45
2,Caderno,Papelaria,Mar,12.5,28
3,Caderno,Papelaria,Abr,12.5,50
4,Lapis,Papelaria,Jan,2.0,80
5,Lapis,Papelaria,Fev,2.0,70
6,Lapis,Papelaria,Mar,2.0,90
7,Lapis,Papelaria,Abr,2.0,65
8,Caneta,Papelaria,Jan,3.5,60
9,Caneta,Papelaria,Fev,3.5,55



Filtro na coluna 'produto'


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45
2,Caderno,Papelaria,Mar,12.5,28
3,Caderno,Papelaria,Abr,12.5,50



Filtro nas colunas 'categoria' e 'produto'


,produto,categoria,mes,preco_unitario,quantidade_vendida
0,Caderno,Papelaria,Jan,12.5,30
1,Caderno,Papelaria,Fev,12.5,45
2,Caderno,Papelaria,Mar,12.5,28
3,Caderno,Papelaria,Abr,12.5,50


## Desafio 3 - Criar colunas

In [ ]:
# df["Total"]
# df["Bonus"]
# df["Lucro"]

# Total = Quantidade * Valor

## Desafio 4 - Estatísticas

In [ ]:
# sum()
# mean()
# max()
# min()
# count()
# value_counts()

## Desafio 5 - Agrupamentos.

In [ ]:
# groupby()
# # Vendedor
# # Total vendido
# # Cidade
# # Quantidade

## Desafio 6 - Ordenação

In [ ]:
# sort_values()

# Top 5 vendedores
# Top 10 produtos
# Maior salário
# Menor salário

## Merge e Pivot
## Desafio 7 - Juntar duas tabelas.

In [ ]:
# merge()

## Desafio 8 - Pivot Table
Transformar: Categoria >> Região

Em uma tabela.

## Desafio 9 - NumPy

In [ ]:
# Criar Arrays
# array()
# zeros()
# ones()
# arange()

## Desafio 10 - Operações
Sem utilizar for e while

In [ ]:
# +
# -
# *
# /
# **

## Desafio 11 - Vetorização
O exercício inteiro será: preco * quantidade

e depois: preco * 1.10

Entender Broadcasting.

# Desafio 12A - Uma função

In [ ]:
# def calcular_total(preco, quantidade):
#     return preco * quantidade

## Desafio 12B - Duas funções.

In [ ]:
# def calcular_total()
# def calcular_desconto()

# total = calcular_total()
# final = calcular_desconto(total)

## Desafio 12C - Três funções

In [ ]:
# def calcular_total()
# def calcular_bonus()
# def classificar()